src/backends/perf_pt/mod.rs

```diff
@@ -0,0 +1,221 @@
+
+// FFI prototypes.
+extern "C" {
+    fn perf_pt_start_tracer(conf: *const PerfPTConf) -> *const c_void;
+    fn perf_pt_stop_tracer(tr_ctx: *const c_void) -> c_int;
+}
+
+// Struct used to communicate a tracing configuration to the C code. Must
+// stay in sync with the C code.
+#[repr(C)]
+struct PerfPTConf {
+    target_pid: pid_t,
+    trace_filename: *const c_char,
+    map_filename: *const c_char,
+    data_bufsize: size_t,
+    aux_bufsize: size_t,
+}
+
+/// A tracer that uses the Linux Perf interface to Intel Processor Trace.
+#[derive(Default)]
+pub struct PerfPTTracer {
+    /// Filename to store the trace to.
+    trace_filename: String,
+    /// PID to trace.
+    target_pid: pid_t,
+    /// Data buffer size, in pages. Must be a power of 2.
+    data_bufsize: size_t,
+    /// Aux buffer size, in pages. Must be a power of 2.
+    aux_bufsize: size_t,
+    /// Opaque C pointer representing the tracer context.
+    tracer_ctx: Option<*const c_void>,
+}
+
+impl PerfPTTracer {
+    /// Create a new tracer.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use traceme::backends::PerfPTTracer;
+    ///
+    /// PerfPTTracer::new().trace_filename("mytrace.ptt").data_bufsize(1024).target_pid(666);
+    /// ```
+    pub fn new() -> Self {
+        Self {
+            trace_filename: String::from("traceme.ptt"),
+            target_pid: unsafe { getpid() },
+            tracer_ctx: None,
+            data_bufsize: 64,
+            aux_bufsize: 1024,
+        }
+    }
+
+    /// Set where to write the trace.
+    ///
+    /// # Arguments
+    ///
+    /// * `filename` - The filename in which to store trace packets.
+    pub fn trace_filename(mut self, filename: &str) -> Self {
+        self.trace_filename = String::from(filename);
+        self
+    }
+
+    /// Select which PID to trace.
+    ///
+    /// By default, the current PID is traced.
+    ///
+    /// # Arguments
+    ///
+    /// * `pid` - The PID to trace.
+    pub fn target_pid(mut self, pid: pid_t) -> Self {
+        self.target_pid = pid;
+        self
+    }
+
+    /// Set the PT data buffer size.
+    ///
+    /// # Arguments
+    ///
+    /// * `size` - The data buffer size to use.
+    pub fn data_bufsize(mut self, size: usize) -> Self {
+        self.data_bufsize = size;
+        self
+    }
+
+    /// Set the PT aux buffer size.
+    ///
+    /// # Arguments
+    ///
+    /// * `size` - The aux buffer size to use.
+    pub fn aux_bufsize(mut self, size: usize) -> Self {
+        self.aux_bufsize = size;
+        self
+    }
+
+    /// Make the map filename by setting/adding a ".map" extension to `trace_filename`.
+    fn make_map_filename(trace_filename: &str) -> Result<String, TraceMeError> {
+        let mut pb = PathBuf::from(trace_filename);
+        if !pb.set_extension("map") {
+            return Err(TraceMeError::InvalidFileName("".to_string()));
+        }
+        let map_filename = pb.to_str().ok_or(TraceMeError::InvalidFileName("".to_string()))?;
```

Comment:
Why do we say there's an InvalidFileName of an empty string?

Reply:
PathBuf::set_extension:
> Returns false and does nothing if self.file_name is None, returns true and updates the extension otherwise.

So if we get `false` there's an empty filename.

We discussed in #13 whether an empty filename should be a distinct error, or an `InvalidFileName` with an empty string, and the latter came up on top:

> ltratt: Personally, I wouldn't even differentiate between empty and invalid filenames: empty filenames are invalid by definition.

https://github.com/softdevteam/traceme/pull/13#issuecomment-350702887

This code will disappear when we use the linker API to get the maps, see #20.

Reply:
OK.